### DISCUSSION TITLE
many to many relationship query with junction table attribute

### DISCUSSION THREAD

**Discussion by ben519** [sqlalchemy/sqlalchemy#7481] [Usage Questions]
[Cross posting here from Stack Overflow](https://stackoverflow.com/questions/70321045/sqlalchemy-orm-many-to-many-relationship-query-with-junction-table-attribute) since my question was well reviewed but didn't receive a proper solution.

I have a many-to-many relationship similar to the one described [here](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object). Notice my Association table includes an `extra_data` field..

```py
class Association(Base):
    __tablename__ = 'association'
    left_id = Column(ForeignKey('left.id'), primary_key=True)
    right_id = Column(ForeignKey('right.id'), primary_key=True)
    extra_data = Column(String(50))

class Parent(Base):
    __tablename__ = 'left'
    id = Column(Integer, primary_key=True)
    children = relationship("Child", secondary="association", back_populates="parents")

class Child(Base):
    __tablename__ = 'right'
    id = Column(Integer, primary_key=True)
    parents = relationship("Parent", secondary="association", back_populates="children")
```

If I want to fetch a particular parent object with its children, I can do

```py
db_parent = db.query(Parent).where(Parent.id == 1).first()
print(db_parent.children[0].id)  # works fine
```

BUT, the `extra_data` field is not included as an attribute of the children.

```py
print(db_parent.children[0].extra_data)
```

> AttributeError: 'Child' object has no attribute 'extra_data'

How can I write fetch the children of a parent such that `extra_data` is included as an attribute? (For tinkering, check out my fully working example at the bottom of my [original post](https://stackoverflow.com/q/70321045/2146894).)

**Comment by zzzeek** [comment]
not sure what you want that to do.  if a certain row in the table called "right" is referenced by 45 or 50 rows in "association", what would be the value of child.extra_data ?  did you want it to be a list of strings?

assuming that can be defined you basically are looking to link these classes together using the [association object pattern](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object) rather than simple many-to-many, and then Child can refer to Child.parent_assocations (a colleciton) and access the .extra_data attributes of each.

**Reply by ben519** [reply]
Let me provide a little more context..

Suppose I have three tables:
- books (*Book*)
- authors (*Author*)
- book_authors (*BookAuthor*)

whereby a book can have multiple authors and an author can have multiple books. Furthermore, each *BookAuthor* has a `blurb` field describing the author's contribution to the book.

Now I'm building an API that lets me fetch a Book. When I fetch the book, I want to return a JSON response like 

```
{
  id: 1,
  title: "My Book",
  authors: [
    {id: 1, name: "Bob", blurb: "Bob is a scientist who wrote chapters 1-3"},
    {id: 2, name: "Sue", blurb: "Sue is an economist who wrote chapter 4"}
  ]
}
```

To generate this JSON object, I start by creating a Pydantic model like

```
class AuthorSchema(BaseModel):
    id: int
    name: str
    blurb: str

class BookSchema(BaseModel):
    id: int
    title: str
    authors: List[AuthorSchema]
```

And I initialize this pydantic model with something like

```
with Session(bind=engine) as session:
    db_book = session.query(Book).where(Book.id == 1).one()
    pd_book = BookSchema.from_orm(db_book)
```

This would work great, except that Pydantic throws an error because `db_book.authors[0]` has no attribute `blurb`. In other words, I need to build an ORM such that the following references are valid:

```
db_book.authors  # returns a collection of authors
db_book.authors[0].id  # returns the id of the first author in the collection
db_book.authors[0].blurb  # returns the blurb about the first author in the collection, with respect to the given book
```

Unfortunately, I don't have control over Pydantic's `from_orm()` constructor. Otherwise I would follow the [Association Object pattern here](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object), in which case the corresponding attributes would be accessible via

```
db_book.authors  # returns a collection of BookAuthors
db_book.authors[0].author.id  # returns the id of the first author in the collection
db_book.authors[0].blurb  # returns the blurb about the first author in the collection, with respect to the given book
```

**Reply by CaselIT** [reply]
You could probably use `column_property` here to manually create a subquery. 
But I'm not sure how you can then modify pydantic to make use of that.
Maybe also an association proxy or a simple `@property` may work if you can make pydantic use them

Regarding your original question `Child` has no `extra_data` since `Child` is not an `Association`. You can have a relationship from child to association that you can use to access the `Association`s connected to a `Child`

**Reply by ben519** [reply]
I appreciate the response and I understand *why* the error occurs, but I don't see how `column_property` can resolve it. (P.S. I'd be happy to pay for your time if either of you are willing to give this a little more attention. Regardless, thanks for your help.)

**Reply by CaselIT** [reply]
> but I don't see how `column_property` can resolve it.

something like this:
```py
class Child(Base):
    ...
    extra_data = column_property(
        select(Association.extra_data)
        .where(Association.right_id == id)
        # TODO: appropriate order by
        .limit(1)
        .correlate_except(Association)
        .scalar_subquery()
    )
```
that said association proxy / property it probably a lot easier if you can make pydantic use that.

> P.S. I'd be happy to pay for your time if either of you are willing to give this a little more attention. Regardless, thanks for your help

there are various ways of sponsoring the project also from here on github if you want to chip in. But that's not a prerequisite to open a discussion or issue :)

**Reply by ben519** [reply]
If I understand correctly, with this solution, when I fetch a random Author with something like `session.query(Author).first()`, it will carry an arbitrary `blurb` attribute. This is quite ugly - an author's `blurb` **only** makes sense within the context of a Book. In other words, blurb shouldn't exist if I simply fetch a random author. `blurb` should only exist if I fetch an author *of a particular book*.

**Reply by CaselIT** [reply]
You are indeed right.

I don't thing you can do what you seek without passing to the Association class explicitly, since the same instance of author should need a different value of blurb depending from which book you access it to.

**Reply by ben519** [reply]
An alternative idea of mine is to create a view only SQL Expression as a mapped attribute of my Book model. Something like the following

```
class Book(Base):
    ...
    
    @property
    def authors(cls):
        s = "SELECT authors.*, book_authors.blurb FROM ... WHERE books.id = self.id"
        return <result of query>
```

I know I can write the SQL to fetch the row set I desire. Any advice on how I can turn my custom SQL query results into a mapped attribute of my `Book` model? I've yet to find an example of this.

**Reply by zzzeek** [reply]
i dont know much about pydantic from_orm but there's nothing special about the "association object" pattern that would suggest other tools even know it's in use.  it just means you have a  class A that has a one to many relationship to assocaitoin class B, and B then has a many-to-one relationship to related class C.

The example you have w/ the JSON:

```
{
  id: 1,
  title: "My Book",
  authors: [
    {id: 1, name: "Bob", blurb: "Bob is a scientist who wrote chapters 1-3"},
    {id: 2, name: "Sue", blurb: "Sue is an economist who wrote chapter 4"}
  ]
}
```

those author rows would be sourced **from the association row**, not from the author row.

without any pydantic details in place (also have you looked at [SQLModel](https://sqlmodel.tiangolo.com/)?), the above JSON to model would be this:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session

Base = declarative_base()


class Book(Base):
    __tablename__ = "book"

    id = Column(Integer, primary_key=True)
    title = Column(String)

    authors = relationship("BookAuthor")


class BookAuthor(Base):
    """this is the assocaition object"""

    __tablename__ = "book_author"
    id = Column(Integer, primary_key=True)
    book_id = Column(ForeignKey("book.id"))
    author_id = Column(ForeignKey("author.id"))

    author = relationship("Author")
    blurb = Column(String)

    @property
    def author_name(self):
        return self.author.author_name


class Author(Base):
    __tablename__ = "author"
    id = Column(Integer, primary_key=True)
    author_name = Column(String)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e) as session:

    bob, sue = Author(author_name="bob"), Author(author_name="sue")

    book = Book(
        title="my book",
        authors=[
            BookAuthor(
                blurb="Bob is a scientist who wrote chapters 1-3", author=bob
            ),
            BookAuthor(
                blurb="Sue is an economist who wrote chapter 4", author=sue
            ),
        ],
    )
    session.add(book)
    session.commit()


with Session(e) as session:
    json = [
        {
            "id": book.id,
            "title": book.title,
            "authors": [
                {"id": book_author.author_id, "name": book_author.author_name}
                for book_author in book.authors
            ],
        }
        for book in session.scalars(
            select(Book).options(
                selectinload(Book.authors).joinedload(BookAuthor.author)
            )
        )
    ]

    print(json)
```

**Reply by ben519** [reply]
@zzzeek this is very helpful - thanks. (Also, I was able to successfully create a read-only `authors` attribute of Book, described [here](https://stackoverflow.com/a/70429015/2146894). But I don't think this is the solution I'll use.)
